File: choir_app_anggota/lib/utils/validators.dart
class Validators {
  Validators._();

  static String? email(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Email wajib diisi';
    }
    final emailRegex = RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$');
    if (!emailRegex.hasMatch(value.trim())) {
      return 'Format email tidak valid';
    }
    return null;
  }

  static String? password(String? value) {
    if (value == null || value.isEmpty) {
      return 'Password wajib diisi';
    }
    if (value.length < 6) {
      return 'Password minimal 6 karakter';
    }
    return null;
  }

  static String? required(String? value, [String fieldName = 'Field']) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName wajib diisi';
    }
    return null;
  }

  static String? phone(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Nomor telepon wajib diisi';
    }
    final phoneRegex = RegExp(r'^[0-9+\-\s]{10,15}$');
    if (!phoneRegex.hasMatch(value.trim())) {
      return 'Format nomor telepon tidak valid';
    }
    return null;
  }
}

File: choir_app_backend/src/modules/anggota/validators.js
/**
 * @module modules/anggota/validators
 * @description Validation chains for anggota endpoints.
 */

const { body, param, query } = require('express-validator');

const createAnggotaValidation = [
  body('nrp').trim().notEmpty().withMessage('NRP wajib diisi'),
  body('nama_lengkap').trim().notEmpty().withMessage('Nama lengkap wajib diisi'),
  body('bagian_suara').isIn(['sopran', 'alto', 'tenor', 'bass']).withMessage('Bagian suara harus sopran, alto, tenor, atau bass'),
  body('email').isEmail().withMessage('Format email tidak valid').normalizeEmail(),
  body('password').isLength({ min: 6 }).withMessage('Password minimal 6 karakter'),
  body('alamat').optional().trim(),
  body('kontak').optional().trim(),
];

const updateAnggotaValidation = [
  param('id').isInt().withMessage('ID harus berupa angka'),
  body('nrp').trim().notEmpty().withMessage('NRP wajib diisi'),
  body('nama_lengkap').trim().notEmpty().withMessage('Nama lengkap wajib diisi'),
  body('bagian_suara').isIn(['sopran', 'alto', 'tenor', 'bass']).withMessage('Bagian suara harus sopran, alto, tenor, atau bass'),
  body('email').isEmail().withMessage('Format email tidak valid').normalizeEmail(),
  body('alamat').optional().trim(),
  body('kontak').optional().trim(),
];

const changePasswordValidation = [
  param('id').isInt().withMessage('ID harus berupa angka'),
  body('oldPassword').notEmpty().withMessage('Password lama wajib diisi'),
  body('newPassword').isLength({ min: 6 }).withMessage('Password baru minimal 6 karakter'),
];

const idParamValidation = [
  param('id').isInt().withMessage('ID harus berupa angka'),
];

const listQueryValidation = [
  query('page').optional().isInt({ min: 1 }).withMessage('Page harus angka positif'),
  query('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit harus 1-100'),
  query('bagian_suara').optional().isIn(['sopran', 'alto', 'tenor', 'bass']).withMessage('Bagian suara tidak valid'),
];

module.exports = {
  createAnggotaValidation,
  updateAnggotaValidation,
  changePasswordValidation,
  idParamValidation,
  listQueryValidation,
};
